[Storage] Custom domain support.

jiacfan opened the pull request:
Support custom domain which targets to solve #9896 and #10520 .

ljian3377 commented:
The `if (parsedUrl.getHost()!.split(".")[1] === "blob")` part and the last `else` do the same thing. Then maybe we should combine them in a same branch?

```ts
      const parsedUrl = URLBuilder.parse(this.url);
      if (parsedUrl.getHost()!.split(".")[1] === "blob") {	      if (parsedUrl.getHost()!.split(".")[1] === "blob") {
        // "https://myaccount.blob.core.windows.net/containername/blob".	        // "https://myaccount.blob.core.windows.net/containername/blob".
        // .getPath() -> /containername/blob	        // .getPath() -> /containername/blob
        const pathComponents = parsedUrl.getPath()!.match("/([^/]*)(/(.*))?");	        const pathComponents = parsedUrl.getPath()!.match("/([^/]*)(/(.*))?");
        containerName = pathComponents![1];	        containerName = pathComponents![1];
        blobName = pathComponents![3];	        blobName = pathComponents![3];
      } else {	      } else if (isIpEndpointStyle(parsedUrl)) {
        // IPv4/IPv6 address hosts... Example - http://192.0.0.10:10001/devstoreaccount1/containername/blob	        // IPv4/IPv6 address hosts... Example - http://192.0.0.10:10001/devstoreaccount1/containername/blob
        // Single word domain without a [dot] in the endpoint... Example - http://localhost:10001/devstoreaccount1/containername/blob	        // Single word domain without a [dot] in the endpoint... Example - http://localhost:10001/devstoreaccount1/containername/blob
        // .getPath() -> /devstoreaccount1/containername/blob	        // .getPath() -> /devstoreaccount1/containername/blob
        const pathComponents = parsedUrl.getPath()!.match("/([^/]*)/([^/]*)(/(.*))?");	        const pathComponents = parsedUrl.getPath()!.match("/([^/]*)/([^/]*)(/(.*))?");
        containerName = pathComponents![2];	        containerName = pathComponents![2];
        blobName = pathComponents![4];	        blobName = pathComponents![4];
      } else {
        // "https://customdomain.com/containername/blob".
        // .getPath() -> /containername/blob
        const pathComponents = parsedUrl.getPath()!.match("/([^/]*)(/(.*))?");
        containerName = pathComponents![1];
        blobName = pathComponents![3];
      }	      }
```

jiacfan commented:
Not combine by design, actually my first code version before PR is combined version, while I changed back to this splitted version for : 1) Reduce regression risk. 2) For ease of maintaining, other wise first branch will be (parsedUrl.getHost()!.split(".")[1] === "blob") || !isIpEndpointStyle(parsedUrl)), which might be not easy for other maintainers.  3) For extensibility, logically this parse logic could be extended further, so we might split existing else further.

---
In reply to: [678883200](https://github.com/Azure/azure-sdk-for-js/pull/10771#issuecomment-678883200) [](ancestors = 678883200)

HarshaNalluru commented on sdk/storage/storage-blob/src/utils/utils.common.ts:
```diff
@@ -541,22 +539,35 @@ export function getAccountNameFromUrl(url: string): string {
   }
 }
 
+export function isIpEndpointStyle(parsedUrl: URLBuilder): boolean {
```
Would you mind adding a couple of unit tests for `isIpEndpointStyle` representing the cases being covered in the regex below?

ljian3377 commented on sdk/storage/storage-blob/src/utils/utils.common.ts:
```diff
@@ -186,19 +186,17 @@ export function extractConnectionStringParts(connectionString: string): Connecti
 
     let accountSas = getValueInConnString(connectionString, "SharedAccessSignature");
-    let accountName = getAccountNameFromUrl(blobEndpoint);
+    let accountName = getValueInConnString(connectionString, "AccountName");
```
This isn't right. This branch is to deal with "SAS connection string" https://github.com/Azure/azure-sdk-for-js/pull/4632 which looks like this "BlobEndpoint=https://myaccount.blob.core.windows.net/;QueueEndpoint=https://myaccount.queue.core.windows.net/;FileEndpoint=https://myaccount.file.core.windows.net/;TableEndpoint=https://myaccount.table.core.windows.net/;SharedAccessSignature=sasString".
I wonder how can we get the account name in this case with custom domain. Or we don't need it in this case?
And why isn't our test case breaking for this?

jiacfan replied:
Thanks for the suggestion!
